source/client-side-encryption/tests/README.rst

```diff
@@ -2188,3 +2188,142 @@ attached to the virtual machine.
+changes:
+
+1. Instead of the standard IMDS TCP endpoint of `169.254.169.254:80`,
+   communicate with the running ``fake_azure`` HTTP server.
+
+2. For each test case, the behavior of the server may be controlled by attaching
+   an additional HTTP header to the sent request: ``X-MongoDB-HTTP-TestParams``.
+
+
+Case 1: Success
+```````````````
+
+Do not set an ``X-MongoDB-HTTP-TestParams`` header.
+
+Upon receiving a response from ``fake_azure``, the driver must decode the
+following information:
+
+1. HTTP status will be ``200 Okay``.
+2. The HTTP body will be a valid JSON string.
+3. The access token will be the string ``"magic-cookie"``.
+4. The expiry duration of the token will be seventy seconds.
+5. The token will have a resource of ``"https://vault.azure.net"``
+
+
+Case 2: Empty JSON
+``````````````````
+
+This case addresses a server returning valid JSON with invalid content.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=empty-json``.
+
+Upon receiving a response:
+
+1. HTTP status will be ``200 Okay``
+2. The HTTP body will be a valid JSON string.
+3. There will be no access token, expiry duration, or resource.
+
+The test case should ensure that this error condition is handled gracefully.
+
+
+Case 3: Bad JSON
+````````````````
+
+This case addresses a server returning malformed JSON.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=bad-json``.
+
+Upon receiving a response:
+
+1. HTTP status will be ``200 Okay``
+2. The response body will contain a malformed JSON string.
+
+The test case should ensure that this error condition is handled gracefully.
+
+
+Case 4: HTTP 404
+````````````````
+
+This case addresses a server returning a "Not Found" response. This is
+documented to occur spuriously within an Azure environment.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=404``.
+
+Upon receiving a response:
+
+1. HTTP status will be ``404 Not Found``.
+2. The response body is unspecified.
+
+The test case should ensure that this error condition is handled gracefully.
+
+
+Case 5: HTTP 500
+````````````````
+
+This case addresses an IMDS server reporting an internal error. This is
+documented to occur spuriously within an Azure environment.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=500``.
+
+Upon receiving a response:
+
+1. HTTP status code will be ``500``.
+2. The response body is unspecified.
+
+The test case should ensure that this error condition is handled gracefully.
+
+
+Case 6: Slow Response
+`````````````````````
+
+This case addresses an IMDS server responding very slowly. Drivers should not
+halt the application waiting on a peer to communicate.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=slow``.
+
+The HTTP response from the ``fake_azure`` server will take at least 1000 seconds
+to complete. The client should report a timeout within a few seconds.
+
+
+Case 7: Large Response
+``````````````````````
+
+This case addresses an IMDS server responding with an abnormally large message.
```

Comment:
after what size a response is considered giant? :) Currently this step fails for me because timeout on receiving server response

Reply:
This is a somewhat arbitrary limit, and may depend on the environment, I suppose. In the C driver I have it 8MB before it errors.

Reply:
does it matter to have this at all here? I suspect that Timeout exception handles this case? I doubt 8MB will be handled quicker than 3/8 seconds?

Reply:
A pathologically broken server could send dozens of MBs to a client within a few seconds. Regardless, this is indeed a pathological test case, and such an error should be handled by the HTTP abstraction, so testing its correctness could be considered "beyond the scope" of IMDS testing, but the same could be said of invalid JSON or a slow server response.

@kevinAlbs what do you think should be considered "in scope" for these tests? I recognize that there's some novelty here that hasn't really been part of testing before.

Reply:
I suggest removing `Case 7: Large Response` if there are difficulties implementing. I assume that is a very unlikely (if possible) failure scenario from an IMDS server.

`Case 6: Slow Response` validates the client-side timeout is correct. I suggest keeping it.

The invalid JSON tests seem like more plausible failure scenarios. If there are no implementation difficulties, I suggest keeping them.